src/async_api/tokio/mod.rs

```diff
@@ -0,0 +1,298 @@
+            is_initialized: AtomicBool::new(false),
+            is_destroyed: AtomicBool::new(false),
+            destroy_signals: Arc::new(Mutex::new(Vec::default())),
+            join_handles: Arc::new(Mutex::new(Vec::default())),
+            driver_join: Arc::new(Mutex::new(None)),
+            driver_receiver: Arc::new(Mutex::new(None)),
+            channel: Arc::new(Mutex::new(Some(channel))),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl SessionHandle for TokioSession {
+    async fn destroy(&self) -> () {
+        self.is_destroyed
+            .store(true, std::sync::atomic::Ordering::SeqCst);
+        let mut guard = self.destroy_signals.lock().await;
+
+        for e in guard.drain(..) {
+            if let Err(e) = e.send(()) {
+                warn!("Unable to send a shutdown signal: {:?}", e);
+            }
+        }
+
+        let mut chan = self.channel.lock().await;
+        chan.take();
+    }
+
+    async fn wait_destroy(&self) -> () {
+        let mut d = self.driver_receiver.lock().await;
+        if let Some(e) = d.take() {
+            let _ = e.await;
+        }
+        ()
+    }
+}
+
+#[async_trait::async_trait]
+impl ActiveSession for TokioSession {
+    fn destroyed(&self) -> bool {
+        self.is_destroyed.load(std::sync::atomic::Ordering::Relaxed)
+    }
+
+    fn initialized(&self) -> bool {
+        self.is_initialized
+            .load(std::sync::atomic::Ordering::Relaxed)
+    }
+
+    async fn initialize(&self, version: &crate::ll::Version) -> () {
+        let mut cfg = self.session_configuration.lock().await;
+        cfg.proto_major = version.major();
+        cfg.proto_minor = version.minor();
+        self.is_initialized
+            .store(true, std::sync::atomic::Ordering::Relaxed);
+    }
+
+    async fn wait_worker_shutdown(&self) -> () {
+        let mut join_handles = self.join_handles.lock().await;
+        for ret in join_all(join_handles.drain(..)).await {
+            if let Err(e) = ret {
+                warn!("Error joining worker of {:?}", e);
+            }
+        }
+    }
+
+    async fn session_configuration(&self) -> Option<super::active_session::SessionConfiguration> {
+        let cfg = self.session_configuration.lock().await;
+        if cfg.proto_major > 0 {
+            Some((*cfg).clone())
+        } else {
+            None
+        }
+    }
+}
+
+struct WorkerIsTerminated(Option<Pin<Box<tokio::sync::oneshot::Receiver<()>>>>);
+impl<'a> Future for &'a mut WorkerIsTerminated {
+    type Output = ();
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let e = self.0.take();
+        if let Some(mut e) = e {
+            if let Poll::Pending = e.as_mut().poll(cx) {
+                self.0 = Some(e);
+                Poll::Pending
+            } else {
+                Poll::Ready(())
+            }
+        } else {
+            Poll::Ready(())
+        }
+    }
+}
```

Comment:
Once you await on a one shot receiver by default it would consume it, which would mean we couldn't poll/check this future more than once since it would be consumed in the first usage. Here we check it, but if its not ready we can just go back to pending. If we ever get a Ready back we won't populate the option back to some, so that its safe to await it again and get the same result (complete) back.

Reply:
Ah, I see that makes sense. Could you add a comment describing what this struct is for? It wasn't obvious to me :)